File: aoc2023/day01/day01.go
package day01

import (
	"aoc2023/common"
	"context"
	"fmt"
	"math"
	"regexp"

	"golang.org/x/sync/errgroup"
)

func solve(ctx context.Context, extractFn func(context.Context, string) (int, error)) (int, error) {
	g, ctx := errgroup.WithContext(ctx)

	lines, readFn := common.LinesFromFile(ctx, "day01/input.txt")
	g.Go(readFn)
	calibrationValues, mapFn := common.MapValues(ctx, lines, extractFn)
	g.Go(mapFn)
	var sum int
	g.Go(common.ForEach(ctx, calibrationValues, func(_ context.Context, v int) error {
		sum += v
		return nil
	}))

	return sum, g.Wait()
}

func SolveA(ctx context.Context) (int, error) {
	return solve(ctx, ExtractCalibrationValue)
}

func SolveB(ctx context.Context) (int, error) {
	return solve(ctx, ExtractCalibrationValueV2)
}

func ExtractCalibrationValue(_ context.Context, line string) (int, error) {
	return extractCalibrationValue(line, calibrationValueExtractorV1)
}

func ExtractCalibrationValueV2(_ context.Context, line string) (int, error) {
	return extractCalibrationValue(line, calibrationValueExtractorV2)
}

func extractCalibrationValue(line string, extractor map[*regexp.Regexp]func(byte) int) (int, error) {
	var first, last int
	firstIndex := math.MaxInt
	lastIndex := math.MinInt
	for regex, convertFn := range extractor {
		matches := regex.FindAllStringIndex(line, -1)

		if len(matches) == 0 {
			continue
		}

		if firstMatch := matches[0]; firstMatch[0] < firstIndex {
			first = convertFn(line[firstMatch[0]])
			firstIndex = firstMatch[0]
		}
		if lastMatch := matches[len(matches)-1]; lastMatch[1] > lastIndex {
			last = convertFn(line[lastMatch[0]])
			lastIndex = lastMatch[1]
		}
	}
	if firstIndex == math.MaxInt || lastIndex == math.MinInt {
		return 0, fmt.Errorf("no digits found in %q", line)
	}
	return first*10 + last, nil
}

var calibrationValueExtractorV1 = map[*regexp.Regexp]func(byte) int{
	regexp.MustCompile(`\d`): func(match byte) int { return int(match - '0') },
}
var calibrationValueExtractorV2 = func() map[*regexp.Regexp]func(byte) int {
	v2 := map[*regexp.Regexp]func(byte) int{
		regexp.MustCompile(`one`):   func(byte) int { return 1 },
		regexp.MustCompile(`two`):   func(byte) int { return 2 },
		regexp.MustCompile(`three`): func(byte) int { return 3 },
		regexp.MustCompile(`four`):  func(byte) int { return 4 },
		regexp.MustCompile(`five`):  func(byte) int { return 5 },
		regexp.MustCompile(`six`):   func(byte) int { return 6 },
		regexp.MustCompile(`seven`): func(byte) int { return 7 },
		regexp.MustCompile(`eight`): func(byte) int { return 8 },
		regexp.MustCompile(`nine`):  func(byte) int { return 9 },
		regexp.MustCompile(`zero`):  func(byte) int { return 0 },
	}
	for k, v := range calibrationValueExtractorV1 {
		v2[k] = v
	}
	return v2
}()

File: aoc2023-rust/src/day01.rs
use crate::common::ascii::parse_digit;
use crate::input_file_lines;
use anyhow::{anyhow, Result};

pub fn solve() -> Result<u32> {
    input_file_lines!("day01.txt")
        .map(|line| line.as_bytes())
        .try_fold(0, |sum, line| {
            let first = first_calibration_value(line)?;
            let last = last_calibration_value(line)?;
            Ok(sum + (first * 10 + last) as u32)
        })
}

const CALIBRATION_TABLE_VALUE: &[(&'static str, u8)] = &[
    ("zero", 0),
    ("one", 1),
    ("two", 2),
    ("three", 3),
    ("four", 4),
    ("five", 5),
    ("six", 6),
    ("seven", 7),
    ("eight", 8),
    ("nine", 9),
];

fn last_calibration_value(line: &[u8]) -> Result<u8> {
    (1..=line.len())
        .rev()
        .map(|i| &line[..i])
        .find_map(|s| {
            s.last()
                .and_then(|&c| parse_digit(c))
                .or_else(|| parse_calibration_value(|cv| s.ends_with(cv)))
        })
        .ok_or_else(|| anyhow!("No value found at end of line"))
}

fn first_calibration_value(line: &[u8]) -> Result<u8> {
    (0..line.len())
        .map(|i| &line[i..])
        .find_map(|s| {
            s.first()
                .and_then(|&c| parse_digit(c))
                .or_else(|| parse_calibration_value(|cv| s.starts_with(cv)))
        })
        .ok_or_else(|| anyhow!("No value found at start of line"))
}

fn parse_calibration_value<F>(match_fn: F) -> Option<u8>
where
    F: Fn(&[u8]) -> bool,
{
    CALIBRATION_TABLE_VALUE
        .iter()
        .find_map(|&(name, value)| match_fn(name.as_bytes()).then(|| value))
}
